### PR TITLE
Initialize cache directory in isolation

### LINKED ISSUE #12167: race condition between creation of `.pytest_cache` and `.pytest_cache/.gitignore`
The solution implemented in https://github.com/pytest-dev/pytest/pull/3982 is not robust - in case test execution is interrupted while a cache element is being written, the `.pytest_cache` directory will exist and be non-empty, but will not contain `.gitignore`.

Because of [this check](https://github.com/pytest-dev/pytest/blob/12e061e2e838fb2c17d54df40a4a11aeb8723c6b/src/_pytest/cacheprovider.py#L192-L193) `.gitignore` ends up never being created.

### PR DESCRIPTION
**Pull request by tamird**
Creating and initializing the cache directory is interruptible; this
avoids a pathological case where interrupting a cache write can cause
the cache directory to never be properly initialized with its supporting
files.

Unify `Cache.mkdir` with `Cache.set` while I'm here so the former also
properly initializes the cache directory.

Closes #12167.

### CODE REVIEW

File: src/_pytest/cacheprovider.py
Comment (nicoddemus) [inline src/_pytest/cacheprovider.py:194]: This code is not *exactly* the same as before, previously if we failed to encode the data to JSON we would not write a file at all, but as is we will have an empty file instead.

Given you mentioned this was for a small performance improvement (which I suspect is not there anyway, would have to benchmark it), let's revert to the previous behavior to ensure we keep the same behavior. 👍

File: src/_pytest/cacheprovider.py
Comment (bluetech) [inline src/_pytest/cacheprovider.py:208]: Not sure how CI passes, but anyway I think we have to keep using `Tuple` and `Union` for a bit

```suggestion
        files: Iterable[Tuple[str, Union[str, bytes]]] = (
```

File: testing/test_cacheprovider.py
Comment (nicoddemus) [inline testing/test_cacheprovider.py:1301]:
```suggestion
    # Original files already existed so left unchanged.
```

File: src/_pytest/cacheprovider.py
Comment (nicoddemus) [inline src/_pytest/cacheprovider.py:222]: I think it is shorter and more explicit to just write the straightforward code:

```suggestion
        path = self._cachedir.joinpath("README.md")
        if not path.is_file():
            path.write_text(README_CONTENT)
        
        path = self._cachedir.joinpath(".gitignore")
        if not path.is_file():
            path.write_text("# Created by pytest automatically.\n*\n")
        
        path = self._cachedir.joinpath("CACHEDIR.TAG")
        if not path.is_file():
            path.write_bytes(CACHEDIR_TAG_CONTENT)
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks a lot @tamird, appreciate the PR!

Left a few comments, please take a look.

Also, please include a changelog entry in `changelog/12167.trivial.rst`:

```
cache: ensure supporting files (``CACHEDIR.TAG``, ``.gitignore``, etc) are always created in the cache directory, even in the event of test session being interrupted.
```

File: src/_pytest/cacheprovider.py
Comment (bluetech) [inline src/_pytest/cacheprovider.py:186]: While it doesn't matter much, it doesn't feel right to me to repeatedly do the `_ensure_cache_dir_and_supporting_files` work on `set`. 

A better solution I think is to do the cache dir creation atomic, i.e. create the dir in a `TemporaryDirectory` (in same dir as the cache dir so it's in the same filesystem) then "commit" it by renaming to `.pytest_cache`.

Comment (nicoddemus) [reply src/_pytest/cacheprovider.py:186]: Hmm that will complicate things a bit if the user is (for some reason) using `cachedir=.` (as in the tests for this).

Comment (nicoddemus) [reply src/_pytest/cacheprovider.py:186]: Actually if the user is using an existing directory for this.

Comment (bluetech) [reply src/_pytest/cacheprovider.py:186]: How does it complicate things? (With the atomic solution the dir is not touched if it already exists)

Comment (nicoddemus) [reply src/_pytest/cacheprovider.py:186]: IIUC we always want to add the `CACHEDIR.TAG`, no?

Comment (tamird) [reply src/_pytest/cacheprovider.py:187]: It's not really clear what we should do when operating in an existing directory. Presumably we don't want to write `.gitignore` because we may end up ignoring the user's files, but we probably do want to write `CACHEDIR.TAG`.

The previous implementation didn't write anything, if the directory already existed. The current implementation tries to write everything, so long as it's not overwriting. Doing all the work in a temporary directory would amount to the previous behavior. What do we want the behavior to be?

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:194]: Done.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py]: Done.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py]: I've actually replaced this code with `open(..., 'x')` to avoid the race between checking existence and opening the file.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py]: It was, but now I need to handle exceptions, so kept the loop.

File: testing/test_cacheprovider.py
Comment (tamird) [reply testing/test_cacheprovider.py]: Done.

File: src/_pytest/cacheprovider.py
Comment (bluetech) [reply src/_pytest/cacheprovider.py:187]: IMO the previous behavior was good, if the directory already exists, don't add the gitignore/CACHEDIR/README.

Comment (tamird) [reply src/_pytest/cacheprovider.py:187]: Done. Note that the solution you originally suggested (creating the directory elsewhere, then moving it) isn't feasible because there's no atomic move operation for directories. In other words, the best we can do is create all the files *before* writing the cache data, greatly reducing the probability of a half-created cache directory but not eliminating it. That's what I've done now.

Comment (nicoddemus) [reply src/_pytest/cacheprovider.py:187]: > IMO the previous behavior was good, if the directory already exists, don't add the gitignore/CACHEDIR/README.

Hmm but isn't this what this PR attempts to fix? The objective is to write the files in case the directory was created by pytest but the files were not written because the session was interrupted suddenly.

If we want the previous behavior, probably best to close the PR at all IMHO.

Comment (tamird) [reply src/_pytest/cacheprovider.py:187]: No, the previous behavior was:

- create the directory
- attempt to write the cache data
- create supporting files

the problem this PR solves is that writing the cache data can take a comparatively long time; by writing supporting files immediately after creating the directory, we narrow the opportunity for corruption significantly.

Comment (tamird) [reply src/_pytest/cacheprovider.py:187]: Sorry, I was mistaken. Indeed, it seems the race between directory creation and initialization of supporting files is significant enough that even when everything happens in sequence we have a problem. That is unless the problem I observed locally arose from `def mkdir` which prior to this PR did not properly initialize the cache directory at all.

Comment (nicoddemus) [reply src/_pytest/cacheprovider.py:187]: Indeed, in the previous version of the code, we did write the supporting files right after creating the directory, and before attempting to write the cache data. 

At first your issue seemed curious to me, because the session would have to be interrupted exactly between `mkdir` and `ensure_supporting_files`, which seems unlikely (but theoretically possible). But ensuring the files are always there seemed OK too.

But if @bluetech thinks the current behavior (only create the files if we create the directory) is OK, then we should probably close this PR instead of introducing these changes (which are OK but do not change the underlying behavior, so I would rather not change anything).

Comment (bluetech) [reply src/_pytest/cacheprovider.py:187]: > there's no atomic move operation for directories

Are you sure? AFAIU, `os.rename` (or `os.replace`) on a directory is atomic. It just renames the directory entry, doesn't move files etc (as long as it's on the same filesystem).

Comment (tamird) [reply src/_pytest/cacheprovider.py:187]: @bluetech I think you're right. The last remaining difficulty is that [`tempfile.TemporaryDirectory`](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) requires you to specify whether it should attempt cleanup at the time of its construction. In our case we want clean to be performed if we're interrupted, but not after we've successfully initialized and moved the directory. I've done the closest thing: ignore cleanup errors. PTAL.

Comment (tamird) [reply src/_pytest/cacheprovider.py:189]: At long last: this is green.

File: src/_pytest/cacheprovider.py
Comment (nicoddemus) [inline src/_pytest/cacheprovider.py:219]: This is not needed, but I think it is good taste for tools to use some prefix in case the directories are lingering for some reason: 

```suggestion
        with tempfile.TemporaryDirectory(prefix="pytest-cache-files-") as d:
```

File: changelog/12167.trivial.rst
Comment (nicoddemus) [inline changelog/12167.trivial.rst:1]: This probably needs to be updated now right?

File: src/_pytest/cacheprovider.py
Comment (nicoddemus) [reply src/_pytest/cacheprovider.py:219]: Also note this comment by @bluetech:

> create the dir in a TemporaryDirectory (in same dir as the cache dir so it's in the same filesystem)

File: src/_pytest/cacheprovider.py
Comment (nicoddemus) [inline src/_pytest/cacheprovider.py:230]: We should handle the a possible race condition here: we do check if `self._cachedir` does not exist at the start of this method, however some other process might have created the directory since then and this will raise an error I believe.

Comment (nicoddemus) [reply src/_pytest/cacheprovider.py:230]: Actually I think we should use `os.rename` explicitly instead of `shutil.move`.

[`os.rename`](https://docs.python.org/3/library/os.html#os.rename):

> On Windows, if dst exists a FileExistsError is always raised. The operation may fail if src and dst are on different filesystems. Use shutil.move() to support moves to a different filesystem.

> On Unix, ... If dst is a non-empty directory, an OSError is raised.

[`shutil.move`](https://docs.python.org/3/library/shutil.html#shutil.move):

> If dst is an existing directory or a symlink to a directory, then src is moved inside that directory. The destination path in that directory must not already exist.

> If dst already exists but is not a directory, it may be overwritten depending on [os.rename()](https://docs.python.org/3/library/os.html#os.rename) semantics.

> If the destination is on the current filesystem, then [os.rename()](https://docs.python.org/3/library/os.html#os.rename) is used. Otherwise, src is copied to the destination using copy_function and then removed. In case of symlinks, a new symlink pointing to the target of src will be created as the destination and src will be removed.

If the directory exists by the time this line is reached, `shutil.move` will move `cachedir` inside the existing directory.

File: changelog/12167.trivial.rst
Comment (tamird) [reply changelog/12167.trivial.rst:1]: Yes, will do.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:219]: I don't think we should do that - we may end up with that directory lingering in the interrupt case. I don't know enough about how signal handlers work in python.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:230]: If it's alright with you, I like the defensive style that's here now.

Comment (tamird) [reply src/_pytest/cacheprovider.py:230]: I'm using shutil.move because it deals with the other-filesystem case. I can try to be more explicit but it'll amount to a partial reimplementation of shuttle.move.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:232]: That argument was added in 3.12 and would do the wrong thing in case writing any of these files fails.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py]: Done.

Comment (nicoddemus) [reply src/_pytest/cacheprovider.py]: Sorry, what's defensive about it, can you clarify?

Comment (tamird) [reply src/_pytest/cacheprovider.py]: It will raise if somehow there's a file at the destination. Obviously this isn't likely today, but this is more resistant to refactoring that might loosen surrounding guarantees. We probably never want to be overwriting here.

**Comment by nicoddemus** [comment]
TBH I'm starting to see this change as -0: we are changing how the supporting files are being created for little reason (as the original reason for the issue is no longer valid). I fear we will be introducing subtle bugs for little gain.

**Comment by tamird** [comment]
> TBH I'm starting to see this change as -0: we are changing how the supporting files are being created for little reason (as the original reason for the issue is no longer valid). I fear we will be introducing subtle bugs for little gain.

The original reason is as valid as it ever was - I saw this symptom on my local machine. It wasn't communicated properly to start, but validity hasn't changed.

File: src/_pytest/cacheprovider.py
Comment (nicoddemus) [reply src/_pytest/cacheprovider.py]: > It will raise if somehow there's a file at the destination. 

Ahh OK due to the `x`/`xt` flags, sorry. Indeed, but we can do similar:

```python
        with tempfile.TemporaryDirectory() as d:
            path = Path(d)
            with open(path.joinpath("README.md"), "xt", encoding="UTF-8") as f:
                f.write_text(README_CONTENT)
            with open(path.joinpath(".gitignore"), "xt", encoding="UTF-8") as f:
                f.write_text("# Created by pytest automatically.\n*\n")
            with open(path.joinpath("CACHEDIR.TAG"), "xb") as f:
                f.write_bytes(CACHEDIR_TAG_CONTENT)        
            shutil.move(path, self._cachedir)
```

My point is that it is more straightforward to just perform the file writing directly, rather than creating a list and iterate over it, with if/else switches based on the data type.

**Comment by nicoddemus** [comment]
I left a few comments which are good to apply, but I will remove myself from reviewer and defer the final decision here to @bluetech. :+1: 

Regardless of the outcome, thanks @tamird for the PR, we appreciate it. :+1:

File: src/_pytest/cacheprovider.py
Comment (jurgenwigg) [inline src/_pytest/cacheprovider.py:229]:
```suggestion
                d.mkdir()
```

It would be good to define how we deal with situation when directory already exists.

File: src/_pytest/cacheprovider.py
Comment (jurgenwigg) [inline src/_pytest/cacheprovider.py:222]:
```suggestion
                d.rename(self._cachedir)
```

File: src/_pytest/cacheprovider.py
Comment (jurgenwigg) [inline src/_pytest/cacheprovider.py:211]: `d` is not a very descriptive name for the value, maybe `temp_dir` would be better?

**Review by jurgenwigg — CHANGES_REQUESTED** [review]

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:211]: Before this change:
```
% git grep 'tempfile.TemporaryDirectory'
doc/en/how-to/fixtures.rst:        with tempfile.TemporaryDirectory() as newpath:
testing/test_assertrewrite.py:                    with tempfile.TemporaryDirectory() as d:
```
I was trying to be consistent. `temp_dir` would be wrong since the type is a string. What would you like?

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:218]: As previously discussed, I prefer the current arrangement which will refuse to overwrite existing files. As for a mapping defined at the top of the file: @nicoddemus had the opposite opinion, so I will leave this as-is.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py]: Done.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py]: We just renamed it on the previous line, so it shouldn't exist. If it does, we should probably allow the exception to bubble up. Done.

File: src/_pytest/cacheprovider.py
Comment (jurgenwigg) [reply src/_pytest/cacheprovider.py:211]: `newpath` looks good :)

File: src/_pytest/cacheprovider.py
Comment (jurgenwigg) [reply src/_pytest/cacheprovider.py:218]: In case when file exists we allow to raise exception which will bubble up and is not matched at all - is it expected behaviour?

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py]: Done.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:218]: Yes. What else would you do? Something unexpected happened.

Comment (jurgenwigg) [reply src/_pytest/cacheprovider.py:218]: Ok

**Review by RonnyPfannschmidt — CHANGES_REQUESTED** [review]
at first glance the proposed mechanism is broken in most linux deployments simply by degrading to copytree on anything that has TMP on tmpfs

File: src/_pytest/cacheprovider.py
Comment (RonnyPfannschmidt) [inline src/_pytest/cacheprovider.py:211]: as i just took note @bluetech already explicitly mentioned we need the TemporaryDirectory in the same directory the cache will be in 

as such the dir parameter would need to be the parent directory and the prefix should be something that hints at the intent

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:224]: Yes, this is a fallback. It is no worse than the current behavior.

File: src/_pytest/cacheprovider.py
Comment (tamird) [reply src/_pytest/cacheprovider.py:211]: Please use existing discussions rather than starting new ones. As I said [here](https://github.com/pytest-dev/pytest/pull/12168#discussion_r1547161910):
> I don't think we should do that - we may end up with that directory lingering in the interrupt case. I don't know enough about how signal handlers work in python.

If you create a temporary directory in the same directory as the cache you may end up leaving the temporary directory there if you are interrupted.

File: src/_pytest/cacheprovider.py
Comment (RonnyPfannschmidt) [reply src/_pytest/cacheprovider.py:224]: my current understanding is that this fallback will happen in all cases on a majority of linux systems existing these days

Comment (tamird) [reply src/_pytest/cacheprovider.py:224]: Not mine! More seriously, please let's discuss in the other thread since the only alternative has been proposed there.

File: src/_pytest/cacheprovider.py
Comment (RonnyPfannschmidt) [reply src/_pytest/cacheprovider.py]: i observe consistent rename failures on my own system when going from the tmpdir to the target


i maintain that if a copytree fallback is consistently needed, its over-complicated and should just write the files directly

Comment (RonnyPfannschmidt) [reply src/_pytest/cacheprovider.py]: it should be noted that it wont fail in the testsuite if pytest itself as the cache dir targets for those would be on the same tmpfs

**Review by bluetech — APPROVED** [review]
LGTM, thanks!

**Review by RonnyPfannschmidt — APPROVED** [review]
given how temporarydirectory actually is used in our case,
a followup at a later point might want to replace it with a simpler mechanism

good work

**Comment by tamird** [comment]
@nicoddemus could you have a look? I believe this is blocked on your approval.

**Comment by nicoddemus** [comment]
It is not blocked, I will defer merging to @bluetech or @RonnyPfannschmidt. 👍

**Comment by tamird** [comment]
Screenshot 2024-04-06 at 13 33 27

**Review by nicoddemus — APPROVED** [review]
Ahh my bad, did not realize that.

Relutanctly approving, as I mentioned I'm -0 on this -- need to mark as "Approve" to pass the merge requirements.

**Comment by tamird** [comment]
Now I just need someone to click merge 😄

**Comment by tamird** [comment]
Turns out the behavior I was seeing which prompted me to send this PR was in fact caused by `mkdir` not properly initializing the cache directory. Plugins (such as [`pytest-insta`](https://github.com/vberlier/pytest-insta/blob/41691fb2d8b14c448dac63e2518e199bfebea0b1/pytest_insta/session.py#L106)) use this function; if they are the first user, then the directory does not get initialized properly. Anyway, this PR fixes that, but we should probably add a test -- I'll send another PR shortly.

**Comment by tamird** [comment]
https://github.com/pytest-dev/pytest/pull/12199

**Pull request merged**
